documentation/docs/configuration.md
---
# !!! DO NOT MODIFY !!!
# This file is auto-generated by the gendoc tool based on the source code of the app.
description: This section describes the configuration parameters and their types for INX-Spammer.
keywords:
- IOTA Node
- Hornet Node
- Spammer
- Configuration
- JSON
- Customize
- Config
- reference
---


# Core Configuration

INX-Spammer uses a JSON standard format as a config file. If you are unsure about JSON syntax, you can find more information in the [official JSON specs](https://www.json.org).

You can change the path of the config file by using the `-c` or `--config` argument while executing `inx-spammer` executable.

For example:
```bash
inx-spammer -c config_defaults.json
```

You can always get the most up-to-date description of the config parameters by running:

```bash
inx-spammer -h --full
```
## <a id="app"></a> 1. Application

| Name                      | Description                                            | Type    | Default value |
| ------------------------- | ------------------------------------------------------ | ------- | ------------- |
| checkForUpdates           | Whether to check for updates of the application or not | boolean | true          |
| [shutdown](#app_shutdown) | Configuration for shutdown                             | object  |               |

### <a id="app_shutdown"></a> Shutdown

| Name                     | Description                                                                                            | Type   | Default value |
| ------------------------ | ------------------------------------------------------------------------------------------------------ | ------ | ------------- |
| stopGracePeriod          | The maximum time to wait for background processes to finish during shutdown before terminating the app | string | "5m"          |
| [log](#app_shutdown_log) | Configuration for log                                                                                  | object |               |

### <a id="app_shutdown_log"></a> Log

| Name     | Description                                         | Type    | Default value  |
| -------- | --------------------------------------------------- | ------- | -------------- |
| enabled  | Whether to store self-shutdown events to a log file | boolean | true           |
| filePath | The file path to the self-shutdown log              | string  | "shutdown.log" |

Example:

```json
  {
    "app": {
      "checkForUpdates": true,
      "shutdown": {
        "stopGracePeriod": "5m",
        "log": {
          "enabled": true,
          "filePath": "shutdown.log"
        }
      }
    }
  }
```

## <a id="logger"></a> 2. Logger

| Name              | Description                                                                 | Type    | Default value |
| ----------------- | --------------------------------------------------------------------------- | ------- | ------------- |
| level             | The minimum enabled logging level                                           | string  | "info"        |
| disableCaller     | Stops annotating logs with the calling function's file name and line number | boolean | true          |
| disableStacktrace | Disables automatic stacktrace capturing                                     | boolean | false         |
| stacktraceLevel   | The level stacktraces are captured and above                                | string  | "panic"       |
| encoding          | The logger's encoding (options: "json", "console")                          | string  | "console"     |
| outputPaths       | A list of URLs, file paths or stdout/stderr to write logging output to      | array   | stdout        |
| disableEvents     | Prevents log messages from being raced as events                            | boolean | true          |

Example:

```json
  {
    "logger": {
      "level": "info",
      "disableCaller": true,
      "disableStacktrace": false,
      "stacktraceLevel": "panic",
      "encoding": "console",
      "outputPaths": [
        "stdout"
      ],
      "disableEvents": true
    }
  }
```

## <a id="inx"></a> 3. INX

| Name                  | Description                                                                                        | Type   | Default value    |
| --------------------- | -------------------------------------------------------------------------------------------------- | ------ | ---------------- |
| address               | The INX address to which to connect to                                                             | string | "localhost:9029" |
| maxConnectionAttempts | The amount of times the connection to INX will be attempted before it fails (1 attempt per second) | uint   | 30               |
| targetNetworkName     | The network name on which the node should operate on (optional)                                    | string | ""               |

Example:

```json
  {
    "inx": {
      "address": "localhost:9029",
      "maxConnectionAttempts": 30,
      "targetNetworkName": ""
    }
  }
```

## <a id="pow"></a> 4. Pow

| Name                | Description                             | Type   | Default value |
| ------------------- | --------------------------------------- | ------ | ------------- |
| refreshTipsInterval | Interval for refreshing tips during PoW | string | "5s"          |

Example:

```json
  {
    "pow": {
      "refreshTipsInterval": "5s"
    }
  }
```

## <a id="restapi"></a> 5. RestAPI

| Name                      | Description                                                                             | Type    | Default value    |
| ------------------------- | --------------------------------------------------------------------------------------- | ------- | ---------------- |
| bindAddress               | The bind address on which the Spammer HTTP server listens                               | string  | "localhost:9092" |
| advertiseAddress          | The address of the Spammer HTTP server which is advertised to the INX Server (optional) | string  | ""               |
| debugRequestLoggerEnabled | Whether the debug logging for requests should be enabled                                | boolean | false            |

Example:

```json
  {
    "restAPI": {
      "bindAddress": "localhost:9092",
      "advertiseAddress": "",
      "debugRequestLoggerEnabled": false
    }
  }
```

## <a id="spammer"></a> 6. Spammer

| Name                                  | Description                                                                        | Type    | Default value                  |
| ------------------------------------- | ---------------------------------------------------------------------------------- | ------- | ------------------------------ |
| autostart                             | Automatically start the spammer on startup                                         | boolean | false                          |
| bpsRateLimit                          | The blocks per second rate limit for the spammer (0 = no limit)                    | float   | 0.0                            |
| cpuMaxUsage                           | Workers remains idle for a while when cpu usage gets over this limit (0 = disable) | float   | 0.8                            |
| workers                               | The amount of parallel running spammers                                            | int     | 0                              |
| message                               | The message to embed within the spam blocks                                        | string  | "We are all made of stardust." |
| tag                                   | The tag of the block                                                               | string  | "HORNET Spammer"               |
| tagSemiLazy                           | The tag of the block if the semi-lazy pool is used (uses "tag" if empty)           | string  | "HORNET Spammer Semi-Lazy"     |
| [valueSpam](#spammer_valuespam)       | Configuration for Value Spam                                                       | object  |                                |
| [tipselection](#spammer_tipselection) | Configuration for tipselection                                                     | object  |                                |

### <a id="spammer_valuespam"></a> Value Spam

| Name               | Description                                                        | Type    | Default value |
| ------------------ | ------------------------------------------------------------------ | ------- | ------------- |
| enabled            | Whether to spam with transaction payloads instead of data payloads | boolean | false         |
| sendBasicOutput    | Whether to send basic outputs                                      | boolean | true          |
| collectBasicOutput | Whether to collect basic outputs                                   | boolean | true          |
| createAlias        | Whether to create aliases                                          | boolean | true          |
| destroyAlias       | Whether to destroy aliases                                         | boolean | true          |
| createFoundry      | Whether to create foundries                                        | boolean | true          |
| destroyFoundry     | Whether to destroy foundries                                       | boolean | true          |
| mintNativeToken    | Whether to mint native tokens                                      | boolean | true          |
| meltNativeToken    | Whether to melt native tokens                                      | boolean | true          |
| createNFT          | Whether to create NFTs                                             | boolean | true          |
| destroyNFT         | Whether to destroy NFTs                                            | boolean | true          |

### <a id="spammer_tipselection"></a> Tipselection

| Name                  | Description                                                                                                 | Type | Default value |
| --------------------- | ----------------------------------------------------------------------------------------------------------- | ---- | ------------- |
| nonLazyTipsThreshold  | The maximum amount of tips in the non-lazy tip-pool before the spammer tries to reduce these (0 = always)   | uint | 0             |
| semiLazyTipsThreshold | The maximum amount of tips in the semi-lazy tip-pool before the spammer tries to reduce these (0 = disable) | uint | 30            |

Example:

```json
  {
    "spammer": {
      "autostart": false,
      "bpsRateLimit": 0,
      "cpuMaxUsage": 0.8,
      "workers": 0,
      "message": "We are all made of stardust.",
      "tag": "HORNET Spammer",
      "tagSemiLazy": "HORNET Spammer Semi-Lazy",
      "valueSpam": {
        "enabled": false,
        "sendBasicOutput": true,
        "collectBasicOutput": true,
        "createAlias": true,
        "destroyAlias": true,
        "createFoundry": true,
        "destroyFoundry": true,
        "mintNativeToken": true,
        "meltNativeToken": true,
        "createNFT": true,
        "destroyNFT": true
      },
      "tipselection": {
        "nonLazyTipsThreshold": 0,
        "semiLazyTipsThreshold": 30
      }
    }
  }
```

## <a id="profiling"></a> 7. Profiling

| Name        | Description                                       | Type    | Default value    |
| ----------- | ------------------------------------------------- | ------- | ---------------- |
| enabled     | Whether the profiling plugin is enabled           | boolean | false            |
| bindAddress | The bind address on which the profiler listens on | string  | "localhost:6060" |

Example:

```json
  {
    "profiling": {
      "enabled": false,
      "bindAddress": "localhost:6060"
    }
  }
```

## <a id="prometheus"></a> 8. Prometheus

| Name            | Description                                                     | Type    | Default value    |
| --------------- | --------------------------------------------------------------- | ------- | ---------------- |
| enabled         | Whether the prometheus plugin is enabled                        | boolean | false            |
| bindAddress     | The bind address on which the Prometheus HTTP server listens on | string  | "localhost:9312" |
| spammerMetrics  | Whether to include the spammer metrics                          | boolean | true             |
| goMetrics       | Whether to include go metrics                                   | boolean | false            |
| processMetrics  | Whether to include process metrics                              | boolean | false            |
| promhttpMetrics | Whether to include promhttp metrics                             | boolean | false            |

Example:

```json
  {
    "prometheus": {
      "enabled": false,
      "bindAddress": "localhost:9312",
      "spammerMetrics": true,
      "goMetrics": false,
      "processMetrics": false,
      "promhttpMetrics": false
    }
  }
```



tools/gendoc/configuration_header.md
---
description: This section describes the configuration parameters and their types for INX-Spammer.
keywords:
- IOTA Node
- Hornet Node
- Spammer
- Configuration
- JSON
- Customize
- Config
- reference
---


# Core Configuration

INX-Spammer uses a JSON standard format as a config file. If you are unsure about JSON syntax, you can find more information in the [official JSON specs](https://www.json.org).

You can change the path of the config file by using the `-c` or `--config` argument while executing `inx-spammer` executable.

For example:
```bash
inx-spammer -c config_defaults.json
```

You can always get the most up-to-date description of the config parameters by running:

```bash
inx-spammer -h --full
```


core/spammer/params.go
package spammer

import (
	"time"

	"github.com/iotaledger/hive.go/core/app"
)

// ParametersSpammer contains the definition of the parameters used by the Spammer.
type ParametersSpammer struct {
	// whether to automatically start the spammer on startup
	Autostart bool `default:"false" usage:"automatically start the spammer on startup"`

	// the blocks per second rate limit for the spammer (0 = no limit)
	BPSRateLimit float64 `name:"bpsRateLimit" default:"0.0" usage:"the blocks per second rate limit for the spammer (0 = no limit)"`
	// workers remains idle for a while when cpu usage gets over this limit (0 = disable)
	CPUMaxUsage float64 `name:"cpuMaxUsage" default:"0.80" usage:"workers remains idle for a while when cpu usage gets over this limit (0 = disable)"`
	// the amount of parallel running spammers
	Workers int `default:"0" usage:"the amount of parallel running spammers"`

	// the message to embed within the spam blocks
	Message string `default:"We are all made of stardust." usage:"the message to embed within the spam blocks"`
	// the tag of the block
	Tag string `default:"HORNET Spammer" usage:"the tag of the block"`
	// the tag of the block if the semi-lazy pool is used (uses "tag" if empty)
	TagSemiLazy string `default:"HORNET Spammer Semi-Lazy" usage:"the tag of the block if the semi-lazy pool is used (uses \"tag\" if empty)"`

	ValueSpam struct {
		// whether to spam with transaction payloads instead of data payloads
		Enabled            bool `default:"false" usage:"whether to spam with transaction payloads instead of data payloads"`
		SendBasicOutput    bool `default:"true" usage:"whether to send basic outputs"`
		CollectBasicOutput bool `default:"true" usage:"whether to collect basic outputs"`
		CreateAlias        bool `default:"true" usage:"whether to create aliases"`
		DestroyAlias       bool `default:"true" usage:"whether to destroy aliases"`
		CreateFoundry      bool `default:"true" usage:"whether to create foundries"`
		DestroyFoundry     bool `default:"true" usage:"whether to destroy foundries"`
		MintNativeToken    bool `default:"true" usage:"whether to mint native tokens"`
		MeltNativeToken    bool `default:"true" usage:"whether to melt native tokens"`
		CreateNFT          bool `default:"true" usage:"whether to create NFTs"`
		DestroyNFT         bool `default:"true" usage:"whether to destroy NFTs"`
	}

	Tipselection struct {
		// NonLazyTipsThreshold is the maximum amount of tips in the non-lazy tip-pool before the spammer tries to reduce these (0 = always).
		// This is used to support the network if someone attacks the tangle by spamming a lot of tips.
		NonLazyTipsThreshold uint32 `default:"0" usage:"the maximum amount of tips in the non-lazy tip-pool before the spammer tries to reduce these (0 = always)"`
		// SemiLazyTipsThreshold is the maximum amount of tips in the semi-lazy tip-pool before the spammer tries to reduce these (0 = disable).
		// This is used to support the network if someone attacks the tangle by spamming a lot of tips.
		SemiLazyTipsThreshold uint32 `default:"30" usage:"the maximum amount of tips in the semi-lazy tip-pool before the spammer tries to reduce these (0 = disable)"`
	}
}

// ParametersRestAPI contains the definition of the parameters used by the Spammer HTTP server.
type ParametersRestAPI struct {
	// BindAddress defines the bind address on which the Spammer HTTP server listens.
	BindAddress string `default:"localhost:9092" usage:"the bind address on which the Spammer HTTP server listens"`
	// AdvertiseAddress defines the address of the Spammer HTTP server which is advertised to the INX Server (optional).
	AdvertiseAddress string `default:"" usage:"the address of the Spammer HTTP server which is advertised to the INX Server (optional)"`
	// DebugRequestLoggerEnabled defines whether the debug logging for requests should be enabled
	DebugRequestLoggerEnabled bool `default:"false" usage:"whether the debug logging for requests should be enabled"`
}

// ParametersPoW contains the definition of the parameters used by PoW.
type ParametersPoW struct {
	// Defines the interval for refreshing tips during PoW.
	RefreshTipsInterval time.Duration `default:"5s" usage:"interval for refreshing tips during PoW"`
}

var ParamsSpammer = &ParametersSpammer{}
var ParamsRestAPI = &ParametersRestAPI{}
var ParamsPoW = &ParametersPoW{}

var params = &app.ComponentParams{
	Params: map[string]any{
		"spammer": ParamsSpammer,
		"restAPI": ParamsRestAPI,
		"pow":     ParamsPoW,
	},
	Masked: nil,
}
